tools/cabana/tools/search.cc

```diff
@@ -0,0 +1,255 @@
+
+    QVBoxLayout *search_results_layout = new QVBoxLayout();
+
+    numberOfSigsLabel = new QLabel(QString("Found: 0"));
+    search_results_layout->addWidget(numberOfSigsLabel);
+
+    data_table = new QTableWidget();
+    data_table->setRowCount(1);
+    data_table->setColumnCount(6);
+
+    search_results_layout->addWidget(data_table);
+
+    main_layout->addLayout(scan_button_layout);
+    main_layout->addLayout(search_parameters_layout);
+    main_layout->addLayout(search_results_layout);
+
+    update();
+
+    QObject::connect(can, &AbstractStream::received, this, &SearchDlg::updateRowData);
+    QObject::connect(can, &AbstractStream::seekedTo, this, &SearchDlg::updateRowData);
+}
+
+void SearchDlg::setRowData(int row, QString msgID, QString bitRange, QString currentValue, QString previousValue){
+    QTableWidgetItem *msg_id = new QTableWidgetItem(msgID);
+    data_table->setItem(row, 0, msg_id);
+
+    QTableWidgetItem *bit_range = new QTableWidgetItem(bitRange);
+    data_table->setItem(row, 1, bit_range);
+
+    QTableWidgetItem *current_value = new QTableWidgetItem(currentValue);
+    data_table->setItem(row, 2, current_value);
+
+    QTableWidgetItem *previous_value = new QTableWidgetItem(previousValue);
+    data_table->setItem(row, 3, previous_value);
+}
+
+void SearchDlg::updateRowData(){
+    data_table->clear();
+    data_table->setRowCount(0);
+
+    if(filteredSignals.size() < 1000){
+        data_table->setRowCount(filteredSignals.size() + 1);
+
+        setRowData(0, QString("Message ID"), QString("Bit Range"), QString("Current Value"), QString("Previous Value"));
+
+        int row=1;
+        for(auto &sig : filteredSignals){
+            setRowData(row, sig.messageID.toString(), QString("%1:%2").arg(sig.offset).arg(sig.offset+sig.size), QString::number(sig.getValue()), QString::number(sig.previousValue));
+            row++;
+        }
+    }
+}
+
+void SearchDlg::update(){
+    first_scan_button->setText(scanningStarted ? "New Scan" : "First Scan");
+    numberOfSigsLabel->setText(QString("Found: ") + QString::number(filteredSignals.size()));
+
+    ScanType selectedValue = (ScanType)(scan_type->currentData().toInt());
+
+    next_scan_button->setEnabled(scanningStarted);
+    undo_scan_button->setEnabled(false);
+
+    scan_type->clear();
+
+    int selectedIndex = -1;
+    int i = 0;
+
+    for(auto scanType : enabledScanTypes()){
+        if(scanType == selectedValue) selectedIndex = i;
+        scan_type->addItem(QString::fromStdString(scanTypeToDisplayName[scanType]), QVariant(scanType));
+        i++;
+    }
+
+    if(selectedIndex != -1){
+        scan_type->setCurrentIndex(selectedIndex);
+    }
+
+    updateRowData();
+
+    for(auto &sig : filteredSignals){
+        sig.previousValue = sig.getValue();
+    }
+}
+
+std::vector<Sig> getAllPossibleSignals(int bits_min, int bits_max){
+    std::vector<Sig> ret;
+    
+    for(auto msg_id : can->last_msgs.keys()) {
```

Comment:
the last_msgs is updated at settings.fps (default is 10hz).  it's mainly used for UI to display the latest signal data. you probably want to use can->rawEvents or can->events to scan through all signals.